### PR TITLE
STY: use Appender pattern for docstrings

### PR DESCRIPTION
**Pull request by jbrockmendel**

### CODE REVIEW

File: statsmodels/base/model.py [hunk 1]
```diff
@@ -877,8 +877,7 @@ def fit(self, start_params=None, method='nm', maxiter=500, full_output=1,
         """
         Fit the model using maximum likelihood.
 
-        The rest of the docstring is from
-        statsmodels.LikelihoodModel.fit
+        See LikelihoodModel.fit.__doc__ for more information.
```
Comment (bashtage) [inline statsmodels/base/model.py]: I wouldn't suggest seeing .__doc__, just LikelihoodModel.fit since LikelihoodModel.fit? will show hep.

**Comment by josef-pkt** [comment]
My first impression after only a superficial look at it: not for some time, if ever

complicated/complex and most likely not stable in pandas, e.g.
https://github.com/pandas-dev/pandas/issues/19932

**Comment by jbrockmendel** [comment]
The pandas discussion you linked is tangential at best.  Nothing there is discussing moving away from the `Appender` pattern.

Totally reasonable for this to be low priority though.

**Comment by bashtage** [comment]
Like like both Appender and Substituter, although I think they would be better named DocAppender and DocSubstituter.  When I was refactoring vector_ar I had to CCP the docstring for seed 4 times.  I would like it much nicer to have a patter like

```
seed_docstring = """
seed : {None, int, RandomState}
    Some description of the inputs
"""

@DocSubstituter(seed=seed_docstring)
def fun(x,y,seed)
"""
    Parameters
    ----------
    x : ndarray
    y: ndarray
    {seed}
"""
```

and the DocSubstituter would take care of leading/training spaces so that `seed_docstring` could have natural format.

I find `@Appender` to be more natural than poking docs in using `func.__doc__ = some_doc`.  Good tools would also allow most of the docstring to live on the function where possible.

**Comment by josef-pkt** [comment]
I think that our current pattern is pretty messy.
However, because switching to something else has widespread consequences, I want to be reasonably sure that we move to something that is appropriate for our case.

Aside:
I rather have `func.__doc__ = some_doc` than a decorator because I can immediately see what it does and in most cases from where the docstring comes. 
A decorator might do lots of extra work and it is not obvious what it does.

(defensive programming: In general I don't trust a-priori decorators, pytest, pip, magic, ..., at least not in our main computational code.)

**Comment by jbrockmendel** [comment]
> A decorator might do lots of extra work and it is not obvious what it does.

Good news!  statsmodels' own decorators can be simplified quite a bit: #5138, #4468

> I want to be reasonably sure that we move to something that is appropriate for our case.

Rule of thumb: "pandas has put more thought effort into X"

**Comment by josef-pkt** [comment]
> Rule of thumb: "pandas has put more thought effort into X"

pandas has more (software) developers, a different use case, and historically I didn't agree with many of the design decisions in pandas. (e.g. pandas had a much too ambiguous, second guessing API for library use)
(my background is scipy development and numpy development as bystander)

**Comment by bashtage** [comment]
I think pandas is a fantastic example of a modern OSS project.  They have both allowed lots of growth and have regularly pruned features -- even widely features --  that get in the way to the goals they have.  
It honestly looks like a "fail fast" type of place.  They are a great example of Say's law in software -- supply (or a useful package) creates its own demand.  They have also aggressively cut the supported versions which helps in CI and code clean up. They are also at less than 1.0, which they take full advantage of.  They have been broadly welcoming of technical refactoring that simplify/unify things (for example, they have a lot of pytest fixtures that have replaced a lot of DIY methods to do things).

It is very different from NumPy which is pretty conservative about API, which makes sense when you are right at the bottom of the stack.  The higher you are, the more rapidly you can fix (and break) your API without causing tons of churn. 

I think projects at the level (in the ecosystem and also version) of statsmodels should resemble pandas more than numpy or scipy (as much as possible).  I think it would be helpful if sm adopted a mantra that there is 1 way to do something and then used it throughout -- this includes allowing wholesale conversion of the old way(s) when a better method has been discovered. In the long run, this approach simplifies changes since if the "new" way turns out to be limiting, it too can be wholesale replaced with the new+ way.

**Comment by jbrockmendel** [comment]
> pandas has more (software) developers

You're right, Teamwork is a big part of the reason this rule of thumb works

> a different use case

Yes, and "docstrings" are in the intersection of those use cases.  Fortunately everyone here is bright enough that pointing out the obvious is not necessary.

**Comment by josef-pkt** [comment]
I don't share that. We are intentionally only using older features of pandas because it is changing too fast and too much experimentation is going on that requires a lot of maintenance work.
Also our developers and contributors are in statistics and econometrics often in applied fields and not datascience developers with a computer science, software development background.

Also it is difficult for us to catch subtle bugs that might be introduced by refactoring, which, I think are in general more difficult for users to catch than getting a wrong DataFrame.
The user base for data handling is much larger than for statistics/econometrics. That is also the case for Stata and R according to user contributed packages #5216.

We are about two thirds of the way to long term replicability and stability e.g. compared to scipy that freezes backwards compatibility almost immediately after a release.
For example, because of our backwards compatibility and relatively good verification of results in unit tests we never had to struggle with getting a reputation of being buggy (except for a few complaints) nor being unreliable and breaking things in every release.

Overall it's a bit of chicken and egg problem, but the potential availability of "eggs" is much smaller in the (applied) statistics/econometrics area than in CS/data science area.
Given the much lower wo/man power for maintenance. it is better to be conservative on widespread refactoring because a lot of time will be spend of getting used to and "fixing" the refactorings.

Example for refactoring that I would like to see:
Every M-Estimator model, including MLE and QMLE models, should define loglike_obs, score_obs, score_factor and hessian_factor (as much as possible).
Every Model should allow for user specified weights (at least var_weights and freq_weights).

But I don't want to experiment with decorators on core computational code like `fit` methods without having looked carefully at the side effects. Nor do I want widespread refactoring that means maintainers and developers have to relearn the code base every year.

Maybe I'm making the point too strongly, we just had large changes to the unit testing code, but that is not the core computational code that users use.

**Comment by jbrockmendel** [comment]
> We are intentionally only using older features of pandas 

Appender and Substitution were derived from matplotlib 1.1.0 seven years ago.

Indistinguishable from "FUD + Reflexive Disagreement"

**Comment by jbrockmendel** [comment]
Just updated most of the rest.  Some questionable cases popped up: approx_hess_cs overwrites a docstring, cov_struct has almost-every class inherit docstrings but misses a few (not clear if this is intentional)

**Comment by jbrockmendel** [comment]
rebased

**Comment by bashtage** [comment]
Did you verify that you can run `python -OO -c "import statsmodels.api as sm"` with this patch?

**Comment by jbrockmendel** [comment]
> Did you verify that you can run python -OO -c "import statsmodels.api as sm" with this patch?

Looks like it still fails, but the problem is in `scipy.stats`

**Comment by jbrockmendel** [comment]
Added a test for -OO import.  Not clear what the right place for such a test is.  Maybe a statsmodels/tests directory?

File: statsmodels/tools/tests/test_decorators.py [hunk 2]
```diff
@@ -40,3 +41,14 @@ def f(self):
         ex.b = -1
 
     assert_equal(ex._cache, dict(b=1,))
+
+
+@pytest.mark.xfail(reason="scipy.stats raises at import-time with -OO")
```
Comment (bashtage) [inline statsmodels/tools/tests/test_decorators.py]: How sure are you that scipy.stats raises?  I tried locally and have seen no issues.

Comment (jbrockmendel) [reply statsmodels/tools/tests/test_decorators.py]: Looks like it's a scipy version issue.  It raises in 1.1.0.  After upgrading to 1.2.1, it no longer raises.  I'll change xfail to xfailif

**Comment by bashtage** [comment]
I get

```
python.exe -OO -c "import statsmodels.api as sm"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\git\statsmodels\statsmodels\api.py", line 24, in <module>
    from .tsa import api as tsa
  File "C:\git\statsmodels\statsmodels\tsa\api.py", line 22, in <module>
    from .ar_model import AR
  File "C:\git\statsmodels\statsmodels\tsa\ar_model.py", line 602, in <module>
    class ARResults(tsbase.TimeSeriesModelResults):
  File "C:\git\statsmodels\statsmodels\tsa\ar_model.py", line 805, in ARResults
    preddoc = AR.predict.__doc__.split('\n')
AttributeError: 'NoneType' object has no attribute 'split'
```

**Comment by jbrockmendel** [comment]
Yep, I now get that too after upgrading numpy.  Got a couple of updates to make.

**Comment by jbrockmendel** [comment]
OK, fixed in ar_model.  After upgrading then downgrading numpy/scipy the errors there went away, so I'll let the CI tell us whether to xfail any dependency versions.

**Comment by bashtage** [comment]
Are these failures recoverable. I think popen needs to use some special options on windows.

**Comment by josef-pkt** [comment]
I don't want any decorator on `fit` methods. (And even less before a release)

Also pandas `_decorator` is a private module, so it might break usage in statsmodels at any time.

**Comment by josef-pkt** [comment]
Simple functions would be enough as alternative when docstrings are modified, AFAIU

`fit.__doc__ = modify( xxx.__doc__, ....)`

**Comment by jbrockmendel** [comment]
> Also pandas _decorator is a private module, so it might break usage in statsmodels at any time.

That's what statsmodels.compat.pandas is for.  The decorator itself IIRC was taken from matplotlib and has been stable for something like 8 years.

**Comment by josef-pkt** [comment]
but given your version of compat pandas, decorator was recently privatized, and so could change at any time without warning.

My dislike of decorators (and of pandas) in core code (like fit) has been roughly stable for 10 years.
(although I'm a bit more accepting towards pandas now)

**Comment by jbrockmendel** [comment]
The “without warning” thing is handled by pandas CI including “downstream” tests.  And the fact that the communities overlap quite a bit.

If you dislike decorators, it’s your project and you can restrict to py2.3 syntax wherever you want.  I’m not yet comfortable with “async” and “await”, so we’re two peas in a pod.

**Comment by bashtage** [comment]
IMO statsmodels should just vendor these functions so that they never break.  I find them to be much more natural to use when compared to 

```
def method(x):
    x += 1
   return x
method.__doc__ = other_method.__doc__ + method_extra % method_dict
```

**Comment by josef-pkt** [comment]
> The “without warning” thing is handled by pandas CI including “downstream” tests. And the fact that the communities overlap quite a bit.

pandas has recently been breaking statsmodels every half a year or so, which is not close to my expected lifetime for software versions.

The question is whether more complex features are worth the increased complexity. I'm fine with cython and would like to get started with using numba, there are huge real and potential gains.
Using doc (or any other) decorators on core computational code does not justify the increase in complexity.

(Like the recent parameterize with dict problem. My original loops were nice and simple. To debug parameterize you have to go through the pytest issues. Those are in unit tests and won't directly affect statsmodels usage.
If I work on fitting problems, then I don't want to worry at the same time about what some decorators might be doing in the background.)

**Comment by bashtage** [comment]
A related issue is that statsmodels should be using a metaclass that handles inherited docstrings.  I use these all over `arch` and `linearmodels` and it saves a ton of time when function signatures are identical but need to be separately implemented for each member.   Some of the uses of `Appender` in the PR would be unnecessary if the metaclass approach was used.

**Comment by bashtage** [comment]
> (Like the recent parameterize with dict problem. My original loops were nice and simple. To debug parameterize you have to go through the pytest issues. Those are in unit tests and won't directly affect statsmodels usage.

All approaches have downsides.  The loop completely hides the source of the failure since you can't observe loop variables.  With a parametrized function, these are trivial to observe.  I would also add that with a good test runner, e.g., PyCharm, you can run right click the failure and re-run only the failed parameterization. 

> Using doc (or any other) decorators on core computational code does not justify the increase in complexity.

Hmmm, last time I looked many parts of `statsmodels` are very challenging to debug given how many times you have to press F11 to get into a function in a wrapped method, or to follow paths that are using some form of introspection (`if hasattr(self, 'something`): do_one_thing(); else: do_somthing_else()`).  You often have to go through 5 or more function calls/dict lookups to actually get to the code that is producing an error. I wouldn't think stepping through a pretty simple decorator would be considered a huge burden given the complexity that is all over `sm`.

**Comment by josef-pkt** [comment]
the problem is cognitive overload
For
`method.__doc__ = other_method.__doc__ + method_extra % method_dict`
it is obvious what it does.

When using a function like
`method.__doc__ = append_substitute(other_method.__doc__, method_extra, method_dict)`
it is obvious that it only does something with the `__doc__` attribute.
The difference is that the decorator operates on the main method, while the function only operates on the doc attribute.


With decorators I have to worry about what they are actually doing, given that they could be doing anything.
Plus, I have to worry how decorators affect inheritance and `super` calls.

Those are too many extra things to worry about when adding a subclass or new fit method, or improve an existing method.

There might be cases where it is fine, like our cache decorator for results attributes, but I don't want to start by adding decorators to the core computational model code.

**Comment by bashtage** [comment]
> the problem is cognitive overload
> For
> `method.__doc__ = other_method.__doc__ + method_extra % method_dict`
> it is obvious what it does.

It is all in the eye of the beholder.  I find 

```
@Appender(doc_string)
def fit(x,y,z):
    """Some basic doc, to be extended"""
    return x+y+z
```

to be much easier to understand thatn 

```
def fit(x,y,x):
    """Some basic doc, to be extended"""
   return x+y+z

fit.__doc__ += doc_string
```

The first example makes is clear where the doc string is when you are near the doc string.  The latter requires you to look down the file.

**Comment by josef-pkt** [comment]
> last time I looked many parts of statsmodels are very challenging to debug

It looks relatively simple to me in many cases, i.e. I usually use `--pdb` on a test case, and it is easy to inspect what the attributes are and walk up and down the.
But sometimes it is very difficult to find in which module, class and method something is defined, which is the downside of the heavy inheritance and delegation that we use.

**Comment by bashtage** [comment]
FWIW scipy also uses a substituter, `@replace_notes_in_docstring`

Also, it appears that `Appender` or `Substituter` are only run once on module import and so that stepping in to a function curing debug doesn't do through the decorator. THis makes sense since it doesn't continually add to .__doc__.

**Comment by bashtage** [comment]
Can you rebase?

**Comment by bashtage** [comment]
Seems to be failing now. Can you take a look. I've started using appender now so this is easier to get in now. It also closes an old rare bug.

**Pull request merged**